### DISCUSSION TITLE
API request slow after upgrading to V9.0.0-rc.82

### DISCUSSION THREAD

**Discussion by Piteraguayo** [directus/directus#6565] [Archive]
### Describe the Bug

After upgrading from rc.50 to rc.82 the request to the API have become 6-7 times long. The "npx directus database migrate:latest" was done successfully. I have 2 servers running: one with rc50 with database non migrated and one with rc.82 and database migrated. For this request:

```
items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"news"}}&limit=3&fields=*.*&sort=-date_created
```

The first server taks 500ms. The second takes almost 7s
When there are lots of requests to the API, the database starts going slowly and collapse.

If I remove the fields level and the filter, the requests become normal in time.

### To Reproduce

Upgrade to 9.0.0-rc.50 doing the "npx directus database migrate:latest". Make a request with 2 fields to a collection with related collections.

### What version of Directus are you using?

V9.0.0-rc.82

### What version of Node.js are you using?

14.17.1

### What database are you using?

Mysql 5.7

### What browser are you using?

Chrome, Firefox

### What operating system are you using?

macOs

### How are you deploying Directus?

Locally

**Comment by rijkvanzanten** [comment]
MySQL 5.x is notoriously slow when it comes to reading from information_schema. I would recommend either using MySQL 8+, or enabling the newly added schema-caching 👍🏻 

That being said, would you be able to share a database dump of this exact schema (no data needed) so we can run some tests about what the slow part is? Either here or through rijk@directus.io)

**Reply by Piteraguayo** [reply]
Hi @rijkvanzanten, thank you for you reply. 
I have to ask for permission to send the dump. I will come back with news.
Thank you again

**Reply by rijkvanzanten** [reply]
Perfect, thank you! 

The main difference between the version you were on and the new one is that Directus now relies on Foreign Key Constraints for its relationships 🤔

**Reply by Piteraguayo** [reply]
I attach the dump file. The API test that is an example of slow performance is the following:

```
items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"news"}}&limit=3&fields=*.*&sort=-date_created
```

As the dump has not data, maybe it is not being slow but with our data on the table "news" with above 50,000 rows, it becomes very slow. 

[develop_NO_DATA_2021-06-30.zip](https://github.com/directus/directus/files/6739486/develop_NO_DATA_2021-06-30.zip)


Thank you

**Reply by rijkvanzanten** [reply]
@Piteraguayo From what I can see, you have news related back to news through `related_posts`. I think you might've run into https://github.com/directus/directus/issues/6184 🤔 How many rows exist in `news_news`?

**Reply by Piteraguayo** [reply]
@rijkvanzanten in `news_news` there are 175 rows. I hope it helps :)

**Reply by rijkvanzanten** [reply]
What about the counts of the other tables related to news? `news_news_categories`, `competitions`, `news_federation_members` etc 🤔

**Reply by Piteraguayo** [reply]
`news` and `news_news_categories`: around 7,000
`news_federation_members`: around 2,000
`news_competions`: around 3,000
`competitions`: 31

That's all :)

Thank you

**Comment by Piteraguayo** [comment]
Hi @rijkvanzanten,
We have done tests with a new MYSQL 8 and with both rc.50 against MYSQL 5 and rc.83 against MYSQL 8 and there is a BIG difference on the DB performance. In short, the results are showing that for the same requests, the rc.83 version is overloading the database 10 times the rc.50 is doing. I don't know if this is being reported by other users.

I attach captures of DB network output for both MYSQL 5 and 8:

MYSQL 5:
![MYSQL_5_DIrectus_rc50](https://user-images.githubusercontent.com/5360439/127311259-72f8733e-733f-4c9a-bc4a-40051ec153e4.png)

MYSQL 8:
![MYSQL_8_Directus_rc83](https://user-images.githubusercontent.com/5360439/127311289-c6809fc3-4d6e-4004-b401-022dec40617a.png)

**Reply by Piteraguayo** [reply]
To be more precise, the first part of the graphic of rc.83 -> MYSQL 8 is with request with 3 levels (`*.*.*`) the second part is with 1 level (`*`). However, the MYSQL 5 -> rc.50, is throwing the more 30MiB with 3 level (`*.*.*`). The difference using 3 levels is from 30MiB to 2.9 GiB. This means 100 times more!!

**Reply by rijkvanzanten** [reply]
Wow! That's not good.. 

> I don't know if this is being reported by other users.

Nope, haven't heard this before. Does your insights show what queries are ran most?

**Reply by Piteraguayo** [reply]
Thank you for your answer @rijkvanzanten 
Here the most queries. All are called in the main page. There is only one with `*.*.*`. On the 3 with `*.*`, "menu" and "config" have a single row and not related tables. The "live_events" with `*.*` has related tables

As I attached the tables structure in previous post, maybe you have any info. 
Thank you again

```
/items/menu?filter[status]=published&fields=*.*
/items/config?fields=*.*
/items/sponsors?filter[status]=published&fields=*
/items/live_events?filter={"status":{"_eq":"published"}}&fields=*.*
/items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"news"}}&limit=3&fields=*&sort=-date_created
/items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"competitions"}}&limit=3&fields=*&sort=-date_created
/items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"athletes"}}&limit=3&fields=*&sort=-date_created
/items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"what_we_do"}}&limit=3&fields=*&sort=-date_created
/items/pages?filter[status]=published&filter[slug][_eq]=home&fields=*
/items/sliders?filter[status]=published&filter[slug][_eq]=home&fields=*.*.*
```

**Reply by rijkvanzanten** [reply]
Thanks! To go even deeper on this, some cloud providers have a way to log what SQL query was executed the most, does your graph above have any such functionality? That would help out tremendously (I'm assuming one of these nested calls and/or filters is doing an extra query somewhere by accident that's causing it to exponentially grow)

**Reply by Piteraguayo** [reply]
`/items/menu?filter[status]=published&fields=*.*`

`/items/config?fields=*.*`

`/items/sponsors?filter[status]=published&fields=*`

`/items/live_events?filter={"status":{"_eq":"published"}}&fields=*.*`

`/items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"news"}}&limit=3&fields=*&sort=-date_created`

`/items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"competitions"}}&limit=3&fields=*&sort=-date_created`

`/items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"athletes"}}&limit=3&fields=*&sort=-date_created`

`/items/news?filter={"status":{"_eq": "published"},"type": {"_contains":"what_we_do"}}&limit=3&fields=*&sort=-date_created`

`/items/pages?filter[status]=published&filter[slug][_eq]=home&fields=*`

`/items/sliders?filter[status]=published&filter[slug][_eq]=home&fields=*.*.*`

**Reply by Piteraguayo** [reply]
I have updated the queries with "code" because it had deleted some fields levels. Sorry

**Reply by Piteraguayo** [reply]
I check if the provider has a tool to know what SQL query was executed the most.

Thank you!!

**Reply by Piteraguayo** [reply]
Hi @rijkvanzanten,
Unfortunately, the provider has not the tool to watch the SQL queries.
I hope you can find an answer about the big difference between versions.

Thank you

**Reply by rijkvanzanten** [reply]
@Piteraguayo Are you around [on Discord](https://directus.chat)? That's a little easier to discuss (and private too 🙂 )

**Reply by Piteraguayo** [reply]
Hi @rijkvanzanten 
Yes, I am on Discord although I don't use it too much. Tell me how we can connect on it, please.
Thank you

**Reply by rijkvanzanten** [reply]
You can DM me on @Rijk#8055 🙂

**Comment by westhouseit** [comment]
Could you enable checks for slow running queries and list any that take longer than a second to complete? Then with those queries an EXPLAIN will help find the reason for slow completion. A common cause is a missing index.

Edit: If you have SSH access then setting up a secure connection and using MySQL Workbench would be helpful.

**Reply by Piteraguayo** [reply]
@westhouseit thank you for your answer. 
It gives me a clue about where to watch. I will come back with my research.

**Comment by Piteraguayo** [comment]
Hi guys, 
we did a change that was not on the front code but in a custom endpoint. This custom endpoint had `*.*.*` request but doing it by ItemsService. Having changed this, the data base out has lowered in a 95%. 
Here the code we had and the code we have (and guess it's the fix at least for our problem), for the case it's useful for you:

**WHAT WE HAD:**

```js
const newInfo = await itemsService.readByQuery({
  sort: [{column: 'title', order: 'asc'}],
  fields: ['*.*.*'],
  filter: {status: {_eq: 'published'}, slug: {_eq: slug}}
});
```


**WHAT WE HAVE:**

```js
const newInfo = await itemsService.readByQuery({
  sort: [{column: 'title', order: 'asc'}],
  fields: [
    '*',
    'related_posts.news_related_id.date_created',
    'related_posts.news_related_id.image',
    'related_posts.news_related_id.slug',
    'related_posts.news_related_id.title',
    'categories.news_categories_id.name'
  ],
  filter: {status: {_eq: 'published'}, slug: {_eq: slug}}
});
```

This said, the mistery still is why it was running well on rc.50 and so slow on rc.83.

Thank you for your attention and I will watch for your news

**Reply by rijkvanzanten** [reply]
> Having changed this, the data base out has lowered in a 95%.

This actually makes a ton of sense, and more or less confirms my suspicion this is caused by https://github.com/directus/directus/issues/6184!

(this gets a tad confusing, bear with me)
By using `*.*.*`, you went from "news" (current) to "related_posts" to "news_related_it", but also back up to "news_current_id" which in turn re-fetches the nested o2m primary keys from "related_posts" by default under the top level news. This combined with the known issue that nested o2ms fetch too much data (see linked posts), would drastically increase your DB out (and node memory usage as well, as all that data coming from the DB is processed in node).

**Reply by Piteraguayo** [reply]
Thank you @rijkvanzanten
So, is it impossible to use 3 levels when we have o2m o m2m?

**Reply by rijkvanzanten** [reply]
No, not impossible. You'll just have to be aware of circular dependencies. When fetching `news -> related_news -> parent_news_item`, you're fetching the parent twice. That combined with bug #6184 means you're fetching way more data than you need. This is not related to the amount of levels you fetch, just due to the fact that you're fetching things with `*` instead of the listed fields 👍🏻

**Comment by Piteraguayo** [comment]
For more information, this custom endpoint has an average of 8-10 requests by minute.
Do you think this could make an output of 2.2GiB on the data base in 5 minutes, with 19 active connections to the data base and finally crash?

**Comment by rijkvanzanten** [comment]
@Piteraguayo Good news! We've figured an optimized query for fetching these relationally nested o2m items, so this perf issue should be resolved starting in rc99 (https://github.com/directus/directus/pull/8944)
